optuna/storages/_rdb/storage.py

```diff
@@ -39,6 +41,43 @@
 _logger = optuna.logging.get_logger(__name__)
 
 
+class _ScopedSessionContextManager:
+    def __init__(self, scoped_session: orm.scoped_session) -> None:
+        self._scoped_session = scoped_session
+
+    def __enter__(self) -> orm.Session:
+        session = self._scoped_session()
+        return session
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[Exception]],
+        exc_value: Optional[Exception],
+        traceback: Optional[TracebackType],
+    ) -> None:
+        try:
+            session = self._scoped_session()
```

Comment:
The `session` object here may or may not refer to the same object as the `session` returned from `__enter__`. E.g. if the caller releases associated connections (by calling `session.remove()`) before entering `__exit__`, the object will refer to a completely new `Session` instance here. Can we somehow make sure to use the same object here? Would holding the `session` as a member attribute work? We might want to instantiate a new `_ScopedSessionContextManager` object in each `with` in that case, instead of holding a single instance as a member.

Reply:
C.f. https://docs.sqlalchemy.org/en/13/orm/contextual.html. By the ways, it is good to know that the session factory is thread safe, in that it creates thread local sessions.

Reply:
> Would holding the session as a member attribute work?
For the first step, I tried it and did not work, because sometimes the member attribute is created by one thread, and is accessed from the other thread.
https://github.com/optuna/optuna/pull/1628/commits/c06a25c2bcb174ca4d4688c8827c29323ff9523d#diff-9aa9e205e3a2038ae9077f72d10abac1R44-R77

Reply:
> We might want to instantiate a new _ScopedSessionContextManager object in each with in that case, instead of holding a single instance as a member.

That is true. I will work on it.

Reply:
To avoid such a situation, I used `@contextmanager` and input `scoped_session` as an argument.

https://github.com/optuna/optuna/pull/1628/commits/3f26b36ea4ee3099ef65a5ad9a54f55c96449490